File: carnival-search-server/src/services/connectors/jira.ts
import axios, { AxiosResponse } from "axios";
import { config } from "../../providers/config";
// import { DB } from "../../db";
// import { AppType, connector_table, document_chunk_table, document_table, DocumentType, SyncStatus } from "../../db/schema";
// import { config } from "../../providers/config";
// import { and, eq, sql } from "drizzle-orm";
// import { text_splitter } from "../../langchain";
// import { Connector } from "../../db_types";
// import { JiraConnectorAdditionalData, JiraConnectorCredentialsData } from "../../types";
// import dayjs from "dayjs";
// import { delete_vector_entity_queue } from "../../providers/queues";
// import { jsonExtract } from "../../utils";
// import { VectorQueueService } from "../vector_queue.service";
// import { Job } from "bullmq";
// import https from "https";

export const fetchJiraAccessToken = async (DTO: { code: string }) => {
    const { code } = DTO;

    const response = await axios.post("https://auth.atlassian.com/oauth/token", {
        grant_type: "authorization_code",
        client_id: config.atlassian_client_id,
        client_secret: config.atlassian_client_secret,
        redirect_uri: config.atlassian_redirect_uri,
        code: code
    }, {
        headers: {
            "Content-Type": "application/json"
        }
    });

    if (response.status !== 200) {
        throw new Error("Failed to fetch Jira access token");
    }

    return {
        access_token: response.data.access_token,
        refresh_token: response.data.refresh_token
    };
};

// const extract_comments_array = (issue: any) => {
//     let comments: string[] = [];

//     // TODO: Extract other types of comments as well, not just text
//     try {
//         for (const comment of issue.fields.comment.comments) {
//             let comment_text: string[] = [];
//             for (const body of comment.body?.content) {
//                 if (body.content) {
//                     for (const item of body.content) {
//                         if (item.type === "text" && item.text?.trim()) {
//                             comment_text.push(item.text)
//                         }
//                     }
//                 }
//             }
//             if (comment_text.length) {
//                 comments.push(comment_text.join(" "))
//             }
//         }

//     } catch (err) {
//         console.log("err --->", err);
//     }

//     return comments;
// }

// export const extract_description_text = (issue: any) => {
//     let description_text: string[] = [];
//     try {
//         if (issue.fields.description) {
//             for (const description of issue.fields.description.content) {
//                 if (description.content) {
//                     for (const item of description.content) {
//                         if (item.type === "text") {
//                             description_text.push(item.text)
//                         }
//                     }
//                 }
//             }
//         }
//     } catch (err) {
//         console.log("err --->", err);
//     }

//     return description_text.join(" ");
// }

// export const initialIndexJira = async (DTO: { connector_id: string, now_date: Date, job: Job }) => {
//     const { connector_id, now_date, job } = DTO;

//     const now_date_formatted = dayjs(now_date).format("YYYY-MM-DD HH:mm");

//     const connector: Connector & { credentials_data: JiraConnectorCredentialsData, additional_data: JiraConnectorAdditionalData } = (await DB.select().from(connector_table).where(eq(connector_table.id, connector_id)))[0] as any;

//     if (!connector) {
//         throw new Error(`Connector with ${connector_id} not found`);
//     }

//     let has_more_issues = true;
//     let next_page_token: string | null = null;

//     while (has_more_issues) {
//         await job.extendLock(job.token ?? "", 60000)
//         let jql = `created < '${now_date_formatted}'`;
//         if (connector.additional_data?.project_ids?.length) {
//             jql += ` AND project IN (${connector.additional_data!.project_ids!.map((project_id: string) => `'${project_id}'`).join(", ")})`;
//         }

//         const response: AxiosResponse<any, any> = await axios.post(`https://api.atlassian.com/ex/jira/${connector.credentials_data.cloud_id}/rest/api/3/search/jql`, {
//             expand: "",
//             fields: [
//                 "*all"
//             ],
//             fieldsByKeys: true,
//             jql,
//             maxResults: 500,
//             nextPageToken: next_page_token ?? undefined,
//         }, {
//             headers: {
//                 Authorization: `Basic ${Buffer.from(`${connector.credentials_data.email}:${connector.credentials_data.access_token}`).toString('base64')}`
//             },
//             httpsAgent: new https.Agent({
//                 keepAlive: true
//             }),
//             timeout: 60000
//         })

//         if (response.data.nextPageToken) {
//             next_page_token = response.data.nextPageToken;
//         } else {
//             has_more_issues = false;
//         }

//         for (const issue of response.data.issues) {
//             const issue_link = `${connector.credentials_data.base_url}/browse/${issue.key}`;

//             const comments_arr = extract_comments_array(issue);
//             const description_text = extract_description_text(issue);
//             let issue_content = `${issue.key} ${issue.fields.summary} ${description_text} \n ${comments_arr.map(comment => `Comment: ${comment}`).join("\n")}`;

//             let metadata: any = {
//                 id: issue.id,
//                 key: issue.key,
//                 created: issue.fields.created,
//                 updated: issue.fields.updated,
//                 status: issue.fields.status?.name,
//                 priority: issue.fields.priority?.name,
//                 reporter: issue.fields.reporter?.displayName,
//                 assignee: issue.fields.assignee?.displayName,
//                 creator: issue.fields.creator?.displayName,
//                 resolution: issue.fields.resolution?.name,
//                 resolution_date: issue.fields.resolutiondate,
//                 labels: issue.fields.project.labels?.length ? issue.fields.project.labels : undefined,
//                 due_date: issue.fields.duedate,
//                 issue_type: issue.fields.issuetype?.name,
//                 project_name: issue.fields.project?.name,
//                 project_key: issue.fields.project?.key,
//             }

//             metadata = Object.fromEntries(Object.entries(metadata).filter(([_, value]) => value !== undefined));

//             issue_content += `\n ${Object.entries(metadata).map(([key, value]) => `${key}: ${value}`).join(", ")}`

//             const document = (await DB.insert(document_table).values({
//                 fk_document_org: connector.fk_connector_org,
//                 fk_document_connector: connector.id,
//                 link: issue_link,
//                 metadata
//             }).returning())[0]

//             const chunks = await text_splitter.splitText(issue_content);

//             await Promise.all(chunks.map(async (chunk, index) => {
//                 await DB.insert(document_chunk_table).values({
//                     fk_chunk_org: connector.fk_connector_org,
//                     fk_chunk_connector: connector.id,
//                     fk_chunk_document: document.id,
//                     chunk_index: index
//                 })

//                 await VectorQueueService.addVectorEntity({
//                     org_id: connector.fk_connector_org,
//                     connector_id: connector.id,
//                     document_id: document.id,
//                     title: `${issue.key} ${issue.fields.summary}`,
//                     chunk_content: chunk,
//                     chunk_index: index
//                 })
//             }))
//         }
//     }
// }

// const deleteDocumentChunks = async (DTO: { document_id: string }) => {
//     const { document_id } = DTO;

//     await DB.delete(document_chunk_table).where(eq(document_chunk_table.fk_chunk_document, document_id));

//     await delete_vector_entity_queue.add("delete_from_document", {
//         document_id
//     })
// }

// export const reindexJira = async (DTO: { connector_id: string, now_date: Date }) => {
//     // const { connector_id, now_date } = DTO;

//     // const connector: Connector & { credentials_data: JiraConnectorCredentialsData } = (await DB.select().from(connector_table).where(eq(connector_table.id, connector_id)))[0] as any;

//     // if (!connector) {
//     //     throw new Error(`Connector with ${connector_id} not found`);
//     // }

//     // const last_synced_at_formatted = dayjs(connector.last_synced_at).format("YYYY-MM-DD HH:mm");
//     // const now_date_formatted = dayjs(now_date).format("YYYY-MM-DD HH:mm");

//     // // Updated issues
//     // let has_more_updated_issues = true;
//     // let next_updated_page_token: string | null = null;

//     // while (has_more_updated_issues) {
//     //     const response: AxiosResponse<any, any> = await axios.post(`${connector.credentials_data.base_url}/rest/api/3/search/jql`, {
//     //         "expand": "",
//     //         "fields": ["*all"],
//     //         "fieldsByKeys": true,
//     //         "jql": `updated >= '${last_synced_at_formatted}' AND updated < '${now_date_formatted}' AND created < '${last_synced_at_formatted}'`,
//     //         "maxResults": 50,
//     //         "nextPageToken": next_updated_page_token ?? undefined
//     //     }, {
//     //         headers: {
//     //             Authorization: `Basic ${Buffer.from(`${connector.credentials_data.email}:${connector.credentials_data.access_token}`).toString('base64')}`
//     //         }
//     //     });

//     //     if (response.data.nextPageToken) {
//     //         next_updated_page_token = response.data.nextPageToken;
//     //     } else {
//     //         has_more_updated_issues = false;
//     //     }

//     //     for (const issue of response.data.issues) {
//     //         const comments_arr = extract_comments_array(issue);
//     //         const description_text = extract_description_text(issue);
//     //         const issue_content = `${issue.key} ${issue.fields.summary} ${description_text}\n${comments_arr.map(comment => `Comment: ${comment}`).join("\n")}`;

//     //         const found_document = (await DB.select().from(document_table)
//     //             .where(
//     //                 and(
//     //                     eq(document_table.fk_document_connector, connector.id),
//     //                     eq(jsonExtract(document_table.metadata, "id"), issue.id)
//     //                 )
//     //             ))[0];

//     //         if (!found_document) {
//     //             throw new Error(`Document not found with issue id ${issue.id}`);
//     //         }

//     //         const metadata = {
//     //             updated: issue.fields.updated,
//     //             statuscategorychangedate: issue.fields.statuscategorychangedate,
//     //             status_id: issue.fields.status.id,
//     //             project_id: issue.fields.project.id,
//     //             assignee_account_id: issue.fields.assignee?.accountId,
//     //             creator_account_id: issue.fields.creator.accountId,
//     //             reporter_account_id: issue.fields.reporter?.accountId,
//     //             issue_type_id: issue.fields.issuetype.id,
//     //         }

//     //         await DB.update(document_table).set({
//     //             title: `${issue.key} - ${issue.fields.summary}`,
//     //             metadata: {
//     //                 ...(found_document.metadata),
//     //                 ...metadata
//     //             }
//     //         }).where(
//     //             eq(document_table.id, found_document.id)
//     //         )

//     //         await deleteDocumentChunks({ document_id: found_document.id });

//     //         const chunks = await text_splitter.splitText(issue_content);

//     //         await Promise.all(chunks.map(async (chunk, index) => {
//     //             await DB.insert(document_chunk_table).values({
//     //                 fk_chunk_org: connector.fk_connector_org,
//     //                 fk_chunk_document: found_document.id,
//     //                 chunk_index: index,
//     //                 content: chunk
//     //             })

//     //             await VectorQueueService.addVectorEntity({
//     //                 org_id: connector.fk_connector_org,
//     //                 document_id: found_document.id,
//     //                 connector_id: connector.id,
//     //                 title: `${issue.key} - ${issue.fields.summary}`
//     //             })
//     //         }))
//     //     }
//     // }


//     // // New issues`;
//     // let has_more_new_issues = true;
//     // let next_new_page_token: string | null = null;

//     // while (has_more_new_issues) {
//     //     const response: AxiosResponse<any, any> = await axios.post(`${connector.credentials_data.base_url}/rest/api/3/search/jql`, {
//     //         "expand": "",
//     //         "fields": ["*all"],
//     //         "fieldsByKeys": true,
//     //         "jql": `created >= "${last_synced_at_formatted}" AND created < "${now_date_formatted}"`,
//     //         "maxResults": 50,
//     //         "nextPageToken": next_new_page_token ?? undefined
//     //     }, {
//     //         headers: {
//     //             Authorization: `Basic ${Buffer.from(`${connector.credentials_data.email}:${connector.credentials_data.access_token}`).toString('base64')}`
//     //         }
//     //     });

//     //     if (response.data.nextPageToken) {
//     //         next_new_page_token = response.data.nextPageToken;
//     //     } else {
//     //         has_more_new_issues = false;
//     //     }

//     //     for (const issue of response.data.issues) {
//     //         const comments_arr = extract_comments_array(issue);
//     //         const description_text = extract_description_text(issue);
//     //         const issue_content = `${issue.key} ${issue.fields.summary} ${description_text}\n${comments_arr.map(comment => `Comment: ${comment}`).join("\n")}`;
//     //         const issue_link = `${connector.credentials_data.base_url}/browse/${issue.key}`;

//     //         const metadata = {
//     //             id: issue.id,
//     //             key: issue.key,
//     //             created: issue.fields.created,
//     //             updated: issue.fields.updated,
//     //             statuscategorychangedate: issue.fields.statuscategorychangedate,
//     //             status_id: issue.fields.status.id,
//     //             project_id: issue.fields.project.id,
//     //             assignee_account_id: issue.fields.assignee?.accountId,
//     //             creator_account_id: issue.fields.creator.accountId,
//     //             reporter_account_id: issue.fields.reporter?.accountId,
//     //             issue_type_id: issue.fields.issuetype.id,
//     //         }

//     //         const document = (await DB.insert(document_table).values({
//     //             fk_document_org: connector.fk_connector_org,
//     //             fk_document_connector: connector.id,
//     //             title: `${issue.key} - ${issue.fields.summary}`,
//     //             link: issue_link,
//     //             type: DocumentType.ISSUE,
//     //             metadata
//     //         }).returning())[0]

//     //         const chunks = await text_splitter.splitText(issue_content);

//     //         await Promise.all(chunks.map(async (chunk, index) => {
//     //             await DB.insert(document_chunk_table).values({
//     //                 fk_chunk_org: connector.fk_connector_org,
//     //                 fk_chunk_document: document.id,
//     //                 chunk_index: index,
//     //                 content: chunk
//     //             })

//     //             /* await VectorQueueService.addVectorEntity({
//     //                 org_id: connector.fk_connector_org,
//     //                 document_id: document.id,
//     //                 connector_id: connector.id,
//     //                 chunk,
//     //                 chunk_index: index
//     //             }) */
//     //         }))
//     //     }
//     // }
// }

// export const getJiraProjects = async (DTO: { base_url: string, email: string, access_token: string }) => {
//     const { base_url, email, access_token } = DTO;

//     const cloud_id_response = await axios.get(`${base_url}/_edge/tenant_info`);
//     const cloud_id = cloud_id_response.data.cloudId;

//     let all_projects: any[] = [];
//     let start_at = 0;
//     let total = 0;

//     do {
//         const response = await axios.get(
//             `https://api.atlassian.com/ex/jira/${cloud_id}/rest/api/3/project/search`,
//             {
//                 headers: {
//                     Authorization: `Basic ${Buffer.from(`${email}:${access_token}`).toString('base64')}`
//                 },
//                 params: {
//                     startAt: start_at,
//                     maxResults: 100,
//                     status: "live"
//                 }
//             }
//         );
//         const { values, total: response_total } = response.data;
//         if (Array.isArray(values)) {
//             all_projects.push(...values);
//         }
//         total = response_total;
//         start_at += 100;
//     } while (all_projects.length < total);

//     return all_projects.map((project: any) => ({
//         label: project.name,
//         value: project.key
//     }));
// }

File: carnival-search-vector-server/tools/jira.py
import requests
import base64
from typing import List, Dict, Any
from type_classes import Document
from utils import formatted_doc
from database import get_db_cursor
import os


def _refresh_jira_token(refresh_token: str):
    try:
        refresh_url = f"https://auth.atlassian.com/oauth/token"
        refresh_data = { 
            "grant_type": "refresh_token", 
            "client_id": os.getenv("ATLASSIAN_CLIENT_ID"), 
            "client_secret": os.getenv("ATLASSIAN_CLIENT_SECRET"), 
            "refresh_token": refresh_token 
        }
        refresh_response = requests.post(refresh_url, data=refresh_data)
        if refresh_response.status_code == 200:
            with get_db_cursor() as cursor:
                cursor.execute(
                    "UPDATE user_connector SET credentials_data = jsonb_set(credentials_data, '{refresh_token}', %s) WHERE credentials_data->>'refresh_token' = %s",
                    (f'"{refresh_response.json().get("refresh_token")}"', refresh_token)
                )
            return refresh_response.json().get("access_token")
        else:
            print(f"Error refreshing Jira token: {refresh_response.text}")
            return None
    except Exception as e:
        print(f"Error refreshing Jira token: {e}")
        return None

def get_jira_issues_by_jql(
    jql: str,
    base_url: str,
    email: str,
    token: str,
    refresh_token: str | None = None,
    limit: int = 20,
    start: int = 0,
    expand: str | None = None
):
    try:
        if refresh_token:
            token = _refresh_jira_token(refresh_token)
            if not token:
                raise ValueError("Could not refresh Jira token")
            headers = {
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        else:
            auth_string = f"{email}:{token}"
            auth_header = base64.b64encode(auth_string.encode("ascii")).decode("ascii")
            headers = {
                "Authorization": f"Basic {auth_header}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
            
        fields = "summary,description,status,assignee,reporter,priority,project,issuetype,updated,resolutiondate"
        tenant_info_url = f"{base_url.rstrip('/')}/_edge/tenant_info"
        tenant_response = requests.get(tenant_info_url)
        tenant_response.raise_for_status()
        cloud_id = tenant_response.json().get("cloudId")
        if not cloud_id:
            raise ValueError("Could not retrieve cloud_id from tenant info")
        url = f"https://api.atlassian.com/ex/jira/{cloud_id}/rest/api/3/search/jql"
        payload = {
            "jql": jql,
            "maxResults": limit
        }
        if fields:
            if isinstance(fields, str):
                if fields == "*all":
                    payload["fields"] = "*all"
                else:
                    payload["fields"] = [f.strip() for f in fields.split(",")]
            else:
                payload["fields"] = fields
        if expand:
            payload["expand"] = expand
        response = requests.post(url, headers=headers, json=payload)
        response.raise_for_status()
        issues = response.json().get("issues", [])
        formatted_response: List[Document] = []
        for issue in issues:
            fields_data = issue.get("fields", {})
            description = fields_data.get("description", "")
            if isinstance(description, dict):
                description = str(description)
            link = f"{base_url.rstrip('/')}/browse/{issue['key']}"
            formatted_response.append(
                formatted_doc(
                    Document(
                        id=issue["key"],
                        title=fields_data.get("summary", issue["key"]),
                        content=description,
                        url=link,
                        metadata={
                            "key": issue["key"],
                            "status": fields_data.get("status", {}).get("name", ""),
                            "assignee": (
                                fields_data.get("assignee", {}).get("displayName", "")
                                if fields_data.get("assignee")
                                else ""
                            ),
                            "reporter": (
                                fields_data.get("reporter", {}).get("displayName", "")
                                if fields_data.get("reporter")
                                else ""
                            ),
                            "priority": (
                                fields_data.get("priority", {}).get("name", "")
                                if fields_data.get("priority")
                                else ""
                            ),
                            "project": fields_data.get("project", {}).get("name", ""),
                            "issue_type": fields_data.get("issuetype", {}).get(
                                "name", ""
                            ),
                            "updated": fields_data.get("updated", ""),
                            "resolutiondate": fields_data.get("resolutiondate", ""),
                        },
                        app_type="jira",
                    )
                )
            )
        return formatted_response
    except Exception as e:
        print(f"Error searching JIRA issues with JQL: {e}")
        return []


def get_jira_users(base_url: str, email: str, token: str, refresh_token: str | None = None, limit: int = 500):
    try:
        if refresh_token:
            token = _refresh_jira_token(refresh_token)
            if not token:
                raise ValueError("Could not refresh Jira token")
            headers = {
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        else:
            auth_string = f"{email}:{token}"
            auth_header = base64.b64encode(auth_string.encode("ascii")).decode("ascii")
            headers = {
                "Authorization": f"Basic {auth_header}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        tenant_info_url = f"{base_url.rstrip('/')}/_edge/tenant_info"
        tenant_response = requests.get(tenant_info_url)
        tenant_response.raise_for_status()
        cloud_id = tenant_response.json().get("cloudId")
        if not cloud_id:
            raise ValueError("Could not retrieve cloud_id from tenant info")
        url = f"https://api.atlassian.com/ex/jira/{cloud_id}/rest/api/2/users/search"
        params = {"username": ".", "maxResults": limit}
        response = requests.get(url, headers=headers, params=params)
        response.raise_for_status()
        users = response.json()
        formatted_response = []
        for user in users:
            formatted_response.append({
                "id": user.get("accountId"),
                "name": user.get("displayName", user.get("name", "")),
            })
        return formatted_response
    except Exception as e:
        print(f"Error fetching Jira users: {e}")
        return []


def get_jira_issue_by_key(issue_key: str, base_url: str, email: str, token: str, refresh_token: str | None = None):
    try:
        if refresh_token:
            token = _refresh_jira_token(refresh_token)
            if not token:
                raise ValueError("Could not refresh Jira token")
            headers = {
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        else:
            auth_string = f"{email}:{token}"
            auth_header = base64.b64encode(auth_string.encode("ascii")).decode("ascii")
            headers = {
                "Authorization": f"Basic {auth_header}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        tenant_info_url = f"{base_url.rstrip('/')}/_edge/tenant_info"
        tenant_response = requests.get(tenant_info_url)
        tenant_response.raise_for_status()
        cloud_id = tenant_response.json().get("cloudId")
        if not cloud_id:
            raise ValueError("Could not retrieve cloud_id from tenant info")
        url = (
            f"https://api.atlassian.com/ex/jira/{cloud_id}/rest/api/2/issue/{issue_key}"
        )
        response = requests.get(url, headers=headers)
        response.raise_for_status()
        issue = response.json()
        fields = issue.get("fields", {})
        description = fields.get("description", "")
        if isinstance(description, dict):
            description = str(description)
        link = f"{base_url.rstrip('/')}/browse/{issue_key}"
        return formatted_doc(
            Document(
                id=issue_key,
                title=fields.get("summary", ""),
                content=description,
                url=link,
                metadata={
                    "key": issue_key,
                    "status": fields.get("status", {}).get("name", ""),
                    "assignee": (
                        fields.get("assignee", {}).get("displayName", "")
                        if fields.get("assignee")
                        else ""
                    ),
                    "reporter": (
                        fields.get("reporter", {}).get("displayName", "")
                        if fields.get("reporter")
                        else ""
                    ),
                    "priority": (
                        fields.get("priority", {}).get("name", "")
                        if fields.get("priority")
                        else ""
                    ),
                    "project": fields.get("project", {}).get("name", ""),
                    "issue_type": fields.get("issuetype", {}).get("name", ""),
                    "updated": fields.get("updated", ""),
                    "resolutiondate": fields.get("resolutiondate", ""),
                },
                app_type="jira",
            )
        )
    except Exception as e:
        print(f"Error fetching JIRA issue by key: {e}")
        return None


def get_jira_projects(base_url: str, email: str, token: str, refresh_token: str | None = None) -> List[Dict[str, Any]]:
    try:
        if refresh_token:
            token = _refresh_jira_token(refresh_token)
            if not token:
                raise ValueError("Could not refresh Jira token")
            headers = {
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        else:
            auth_string = f"{email}:{token}"
            auth_header = base64.b64encode(auth_string.encode("ascii")).decode("ascii")
            headers = {
                "Authorization": f"Basic {auth_header}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        tenant_info_url = f"{base_url.rstrip('/')}/_edge/tenant_info"
        tenant_response = requests.get(tenant_info_url)
        tenant_response.raise_for_status()
        cloud_id = tenant_response.json().get("cloudId")
        if not cloud_id:
            raise ValueError("Could not retrieve cloud_id from tenant info")

        url = f"https://api.atlassian.com/ex/jira/{cloud_id}/rest/api/2/project"

        response = requests.get(url, headers=headers)
        response.raise_for_status()

        projects = response.json()

        filtered_projects = [
            {
                "id": project.get("id"),
                "key": project.get("key"),
                "name": project.get("name"),
            }
            for project in projects
        ]

        return filtered_projects

    except requests.exceptions.HTTPError as http_err:
        print(f"HTTP error occurred while getting JIRA projects: {http_err}")
        print(f"Response body: {http_err.response.text}")
        return []
    except Exception as e:
        print(f"An error occurred while getting JIRA projects: {e}")
        return []


def get_all_jira_fields(
    base_url: str, email: str, token: str, refresh_token: str | None = None
) -> List[Dict[str, Any]]:
    try:
        if refresh_token:
            token = _refresh_jira_token(refresh_token)
            if not token:
                raise ValueError("Could not refresh Jira token")
            headers = {
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        else:
            auth_string = f"{email}:{token}"
            auth_header = base64.b64encode(auth_string.encode("ascii")).decode("ascii")
            headers = {
                "Authorization": f"Basic {auth_header}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        tenant_info_url = f"{base_url.rstrip('/')}/_edge/tenant_info"
        tenant_response = requests.get(tenant_info_url)
        tenant_response.raise_for_status()
        cloud_id = tenant_response.json().get("cloudId")
        if not cloud_id:
            raise ValueError("Could not retrieve cloud_id from tenant info")

        url = f"https://api.atlassian.com/ex/jira/{cloud_id}/rest/api/2/field"

        response = requests.get(url, headers=headers)
        response.raise_for_status()

        all_fields = response.json()

        fields = [
            {
                "id": field.get("id"),
                "key": field.get("key"),
                "name": field.get("name"),
                "type": field.get("schema", {}).get("type"),
            }
            for field in all_fields
        ]

        return fields

    except requests.exceptions.HTTPError as http_err:
        print(f"HTTP error occurred while getting JIRA fields: {http_err}")
        print(f"Response body: {http_err.response.text}")
        return []
    except Exception as e:
        print(f"An error occurred while getting JIRA fields: {e}")
        return []


def get_jira_issue_statuses(base_url: str, email: str, token: str, refresh_token: str | None = None) -> List[str]:
    """Get a list of all available issue statuses in JIRA."""
    try:
        if refresh_token:
            token = _refresh_jira_token(refresh_token)
            if not token:
                raise ValueError("Could not refresh Jira token")
            headers = {
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        else:
            auth_string = f"{email}:{token}"
            auth_header = base64.b64encode(auth_string.encode("ascii")).decode("ascii")
            headers = {
                "Authorization": f"Basic {auth_header}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        tenant_info_url = f"{base_url.rstrip('/')}/_edge/tenant_info"
        tenant_response = requests.get(tenant_info_url)
        tenant_response.raise_for_status()
        cloud_id = tenant_response.json().get("cloudId")
        if not cloud_id:
            raise ValueError("Could not retrieve cloud_id from tenant info")

        url = f"https://api.atlassian.com/ex/jira/{cloud_id}/rest/api/2/status"

        response = requests.get(url, headers=headers)
        response.raise_for_status()

        statuses = response.json()

        status_names = [status.get("name") for status in statuses if status.get("name")]

        return list(set(status_names))

    except requests.exceptions.HTTPError as http_err:
        print(f"HTTP error occurred while getting JIRA statuses: {http_err}")
        print(f"Response body: {http_err.response.text}")
        return []
    except Exception as e:
        print(f"An error occurred while getting JIRA statuses: {e}")
        return []


def get_jira_issue_priorities(base_url: str, email: str, token: str, refresh_token: str | None = None) -> List[str]:
    """Get a list of all available issue priorities in JIRA."""
    try:
        if refresh_token:
            token = _refresh_jira_token(refresh_token)
            if not token:
                raise ValueError("Could not refresh Jira token")
            headers = {
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        else:
            auth_string = f"{email}:{token}"
            auth_header = base64.b64encode(auth_string.encode("ascii")).decode("ascii")
            headers = {
                "Authorization": f"Basic {auth_header}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        tenant_info_url = f"{base_url.rstrip('/')}/_edge/tenant_info"
        tenant_response = requests.get(tenant_info_url)
        tenant_response.raise_for_status()
        cloud_id = tenant_response.json().get("cloudId")
        if not cloud_id:
            raise ValueError("Could not retrieve cloud_id from tenant info")

        url = f"https://api.atlassian.com/ex/jira/{cloud_id}/rest/api/2/priority"

        response = requests.get(url, headers=headers)
        response.raise_for_status()

        priorities = response.json()

        # Extract priority names and return as list of strings
        priority_names = [priority.get("name") for priority in priorities if priority.get("name")]

        return priority_names

    except requests.exceptions.HTTPError as http_err:
        print(f"HTTP error occurred while getting JIRA priorities: {http_err}")
        print(f"Response body: {http_err.response.text}")
        return []
    except Exception as e:
        print(f"An error occurred while getting JIRA priorities: {e}")
        return []


def get_jira_issue_types(base_url: str, email: str, token: str, refresh_token: str | None = None) -> List[str]:
    """Get a list of all available issue types in JIRA."""
    try:
        if refresh_token:
            token = _refresh_jira_token(refresh_token)
            if not token:
                raise ValueError("Could not refresh Jira token")
            headers = {
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        else:
            auth_string = f"{email}:{token}"
            auth_header = base64.b64encode(auth_string.encode("ascii")).decode("ascii")
            headers = {
                "Authorization": f"Basic {auth_header}",
                "Content-Type": "application/json",
                "Accept": "application/json",
            }
        tenant_info_url = f"{base_url.rstrip('/')}/_edge/tenant_info"
        tenant_response = requests.get(tenant_info_url)
        tenant_response.raise_for_status()
        cloud_id = tenant_response.json().get("cloudId")
        if not cloud_id:
            raise ValueError("Could not retrieve cloud_id from tenant info")

        url = f"https://api.atlassian.com/ex/jira/{cloud_id}/rest/api/2/issuetype"

        response = requests.get(url, headers=headers)
        response.raise_for_status()

        issue_types = response.json()

        # Extract issue type names and return as list of strings
        type_names = [issue_type.get("name") for issue_type in issue_types if issue_type.get("name")]

        return type_names

    except requests.exceptions.HTTPError as http_err:
        print(f"HTTP error occurred while getting JIRA issue types: {http_err}")
        print(f"Response body: {http_err.response.text}")
        return []
    except Exception as e:
        print(f"An error occurred while getting JIRA issue types: {e}")
        return []

